### PR TITLE
[WC-2647]: Fix AutoNumber formatter

### PR DESCRIPTION
**Pull request by rahmanunver**
### Pull request type

Bug fix (non-breaking change which fixes an issue)

---

### Description
Fixing filters not working for AutoNumber, WIP for falsy values and tests for overriden formatter.

### CODE REVIEW

File: packages/shared/widget-plugin-filtering/src/stores/NumberInputFilterStore.ts [hunk 1]
```diff
@@ -1,32 +1,36 @@
...
+        let { formatter } = attributes[0];
+        formatter = formatterFix(formatter);
         super(new NumberArgument(formatter), new NumberArgument(formatter), "equal", attributes);
         makeObservable(this, {
             updateProps: action,
             fromJSON: action,
             fromViewState: action
         });
-
+        this.formatter = formatter;
```
Comment (iobuhov) [inline packages/shared/widget-plugin-filtering/src/stores/NumberInputFilterStore.ts:33]: iI think you should not set formatter on FilterStore. It should be used by NumberArgument internally.

File: packages/shared/widget-plugin-filtering/src/stores/NumberInputFilterStore.ts [hunk 2]
```diff
@@ -36,9 +40,8 @@ export class NumberInputFilterStore
         if (!comparer.shallow(this._attributes, attributes)) {
             this._attributes = attributes;
         }
-        const formatter = attributes.at(0)?.formatter;
-        this.arg1.updateProps(formatter as NumberFormatter);
-        this.arg2.updateProps(formatter as NumberFormatter);
+        this.arg1.updateProps(this.formatter as NumberFormatter);
+        this.arg2.updateProps(this.formatter as NumberFormatter);
```
Comment (iobuhov) [inline packages/shared/widget-plugin-filtering/src/stores/NumberInputFilterStore.ts:44]: I feel like we can ignore updating formatter. So you can just remove the whole thing.

Comment (iobuhov) [reply packages/shared/widget-plugin-filtering/src/stores/NumberInputFilterStore.ts:44]: And actually it's a good question that you can ask wtf, weather it's changes or it's safe to keep this object.

**Review by iobuhov — CHANGES_REQUESTED** [review]

File: packages/shared/widget-plugin-filtering/src/stores/NumberInputFilterStore.ts (hunk 2 above)
Comment (rahmanunver) [reply packages/shared/widget-plugin-filtering/src/stores/NumberInputFilterStore.ts:44]: Got in touch, first impression no changes apart from the top of the props assigning initially, I'd say a valid suggestion from your side.

File: packages/shared/widget-plugin-filtering/src/stores/NumberInputFilterStore.ts (hunk 1 above)
Comment (rahmanunver) [reply packages/shared/widget-plugin-filtering/src/stores/NumberInputFilterStore.ts]: made the change

File: packages/shared/widget-plugin-filtering/src/stores/NumberInputFilterStore.ts (hunk 2 above)
Comment (rahmanunver) [reply packages/shared/widget-plugin-filtering/src/stores/NumberInputFilterStore.ts]: removed the formatter updates

File: packages/shared/widget-plugin-filtering/src/stores/Argument.ts [hunk 3]
```diff
@@ -53,9 +53,10 @@ export class ArgumentBase<T extends Big | Date | string> {
     set displayValue(textValue: string) {
         this._lastSetTextValue = textValue;
         const parseResult = this._formatter?.parse(this._lastSetTextValue);
-        this.isValid = parseResult.valid;
-        if (parseResult.valid && !this.equals(this._value, parseResult.value)) {
-            this._value = parseResult.value;
+
+        this.isValid = parseResult.valid || (parseResult as any).type === "AutoNumber";
+        if ((parseResult as any).valid || (this.isValid && !this.equals(this._value, (parseResult as any).value))) {
+            this._value = (parseResult as any).value;
```
Comment (iobuhov) [inline packages/shared/widget-plugin-filtering/src/stores/Argument.ts:59]: Why you changing this part? Formatter should work as before I think.

Comment (rahmanunver) [reply packages/shared/widget-plugin-filtering/src/stores/Argument.ts:59]: If the formatter gets the value as `""` then a `new Big(value)` isn't created, which falls into catch block, eventually leaving the empty filter value unset. Looking for another way to handle this.

Comment (iobuhov) [reply packages/shared/widget-plugin-filtering/src/stores/Argument.ts:59]: Then we have to fix this in formatter implementation. Just check what is returned from other formatter (Integer e.g.) and do same in the fix

**Review by iobuhov — APPROVED** [review]

**Review by leonardomendix — APPROVED** [review]

**Pull request merged**